### PR TITLE
Migrate access control

### PR DESCRIPTION
**Pull request by yotamdaniel**

### CODE REVIEW

File: contracts/AccessController.sol [hunk 1]
```diff
@@ -2,129 +2,88 @@
...
 /// @dev Contract used to inherit standard role enforcement across Archimedes contracts
-abstract contract AccessController is AccessControl, ReentrancyGuard {
+abstract contract AccessController is AccessControlUpgradeable {
     bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
     bytes32 public constant EXECUTIVE_ROLE = keccak256("EXECUTIVE_ROLE");
     bytes32 public constant GOVERNOR_ROLE = keccak256("GOVERNOR_ROLE");
     bytes32 public constant GUARDIAN_ROLE = keccak256("GUARDIAN_ROLE");
 
-    address private _addressAdmin;
+    // address private _addressDefaultAdmin;
```
Comment (yotamdaniel) [inline contracts/AccessController.sol:15]: Remove

File: contracts/AccessController.sol [hunk 2]
```diff
@@ -2,129 +2,88 @@
...
+    // address private _addressAdmin;
     address private _addressExecutive;
     address private _addressGovernor;
     address private _addressGuardian;
 
-    bool private _initialized;
-    bool private _initializing;
-    bool private _rolesSet;
+    // modifier onlyDefaultAdmin() {
```
Comment (yotamdaniel) [inline contracts/AccessController.sol:21]: remove

File: contracts/CDPosition.sol [hunk 3]
```diff
@@ -3,6 +3,7 @@ pragma solidity 0.8.13;
 
 import "hardhat/console.sol";
 import {AccessController} from "./AccessController.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
```
Comment (yotamdaniel) [inline contracts/CDPosition.sol:6]: Try to remove

File: contracts/CDPosition.sol [hunk 4]
```diff
@@ -139,4 +138,11 @@ contract CDPosition is AccessController {
     function getShares(uint256 nftID) external view nftIDMustExist(nftID) returns (uint256) {
         return _nftCDP[nftID].shares;
     }
+
+    function initialize(address admin) public initializer {
+        _grantRole(ADMIN_ROLE, admin);
```
Comment (yotamdaniel) [inline contracts/CDPosition.sol:143]: Need to change to msgSender()  or just not set other roles?

Comment (yotamdaniel) [reply contracts/CDPosition.sol:143]: Check other contracts for best practice

File: contracts/Exchanger.sol [hunk 5]
```diff
@@ -55,15 +58,15 @@ contract Exchanger is IExchanger, AccessController {
      * @param addressPoolLvUSD3CRV 3CRV+LvUSD pool address
      * @param addressPoolOUSD3CRV 3CRV+OUSD pool address
      */
-    function init(
+    function setDependencies(
```
Comment (yotamdaniel) [inline contracts/Exchanger.sol:61]: Add to runbook that we need to call this method before using contracts

File: contracts/Mocks/AccessControllerMock.sol [hunk 6]
```diff
@@ -4,17 +4,17 @@ pragma solidity 0.8.13;
 import "hardhat/console.sol";
```
Comment (yotamdaniel) [inline contracts/Mocks/AccessControllerMock.sol:4]: Try removing this file

File: contracts/PositionToken.sol [hunk 7]
```diff
@@ -3,46 +3,65 @@ pragma solidity 0.8.13;
...
-    Counters.Counter private _positionTokenIdCounter;
+    CountersUpgradeable.Counter private _positionTokenIdCounter;
 
     event NFTCreated(uint256 indexed _positionId, address indexed _minter);
     event NFTBurned(uint256 indexed _positionId, address indexed _redeemer);
 
-    constructor(address admin) ERC721("PositionToken", "PNT") AccessController(admin) {}
-
-    function init() external initializer onlyAdmin {}
+    // constructor(address admin) ERC721("PositionToken", "PNT") {}
```
Comment (yotamdaniel) [inline contracts/PositionToken.sol:26]: remove

File: contracts/PositionToken.sol [hunk 8]
```diff
@@ -3,46 +3,65 @@ pragma solidity 0.8.13;
...
         return positionTokenId;
     }
 
-    function exists(uint256 positionTokenId) external view expectInitialized returns (bool) {
+    function exists(uint256 positionTokenId) external view returns (bool) {
         return _exists(positionTokenId);
     }
 
+    function initialize(address admin) public initializer {
+        __ERC721_init("PositionToken", "PNT");
```
Comment (yotamdaniel) [inline contracts/PositionToken.sol:43]: Change to ArchimededPositionToken , APT

File: contracts/VaultOUSD.sol [hunk 9]
```diff
@@ -1,36 +1,30 @@
...
 
 import {ERC20} from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
-import {IERC20Metadata} from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+import {IERC20MetadataUpgradeable} from "@openzeppelin/contracts-upgradeable/interfaces/IERC20MetadataUpgradeable.sol";
 import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
-import {ERC4626} from "./standard/ERC4626.sol";
 import {AccessController} from "./AccessController.sol";
 import {ParameterStore} from "./ParameterStore.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
```
Comment (yotamdaniel) [inline contracts/VaultOUSD.sol:10]: remove?

File: package.json [hunk 10]
```diff
@@ -17,7 +17,7 @@
         "test:ci": "npm run clean && npm run typechain && npx hardhat test --parallel",
         "console": "ts-node console.ts",
         "test:full": "hardhat typechain && hardhat test && hardhat test integrationTests/iArchToken.ts",
-        "test": "hardhat typechain && hardhat test --parallel",
+        "test": "hardhat typechain && hardhat test",
```
Comment (yotamdaniel) [inline package.json:20]: Make sure github uses test:Ci

File: test/ContractTestContext.ts [hunk 11]
```diff
@@ -84,6 +69,13 @@ export type ContractTestContext = ArchContracts & {
     // Archimedes contracts
     // TODO - how to make this type of parameterStore? Failing when I just set it :(
     parameterStore: Contract;
+    cdp: Contract;
+    coordinator: Contract;
+    exchanger: Contract;
+    // leverageAllocator: LeverageAllocator;
```
Comment (yotamdaniel) [inline test/ContractTestContext.ts:75]: remove

File: test/ParameterStore.ts [hunk 12]
```diff
@@ -203,13 +203,12 @@ describe("ParameterStore Access Control tests", async function () {
...
         parameterStore = r.parameterStore;
-        parameterStore.addGovernor(r.addr1.address);
-        parameterStore.revokeGovernor(r.owner.address);
+        parameterStore.setGovernor(r.addr1.address);
     });
 
     it("Should not be able to change governor if not admin", async function () {
-        const changePromise = parameterStore.connect(r.addr2).addGovernor(r.addr3.address);
-        await expect(changePromise).to.be.revertedWith("Caller is not an Admin");
+        const changePromise = parameterStore.connect(r.addr2).x(r.addr3.address);
```
Comment (yotamdaniel) [inline test/ParameterStore.ts:210]: Changer tp setGov

File: package.json (hunk 10 above)
Comment (yotamdaniel) [reply package.json:20]: Yes, github uses test:ci

File: contracts/Coordinator.sol [hunk 13]
```diff
@@ -12,13 +12,16 @@ import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol
 import {ReentrancyGuard} from "@openzeppelin/contracts/security/ReentrancyGuard.sol";
 import {AccessController} from "./AccessController.sol";
 
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
```
Comment (jasonstockman) [inline contracts/Coordinator.sol:16]: duplicate of 12?

File: contracts/Coordinator.sol [hunk 14]
```diff
@@ -38,16 +41,14 @@ contract Coordinator is ICoordinator, AccessController {
         _;
     }
 
-    constructor(address admin) AccessController(admin) {}
-
-    function init(
+    function setDependencies(
```
Comment (jasonstockman) [inline contracts/Coordinator.sol:44]: we arent using the "initialize()" pattern?

File: contracts/Exchanger.sol [hunk 15]
```diff
@@ -43,7 +46,7 @@ contract Exchanger is IExchanger, AccessController {
      */
     uint256 internal _curveGuardPercentage;
 
-    constructor(address admin) AccessController(admin) {}
+    // constructor(address admin) AccessController(admin) {}
```
Comment (jasonstockman) [inline contracts/Exchanger.sol:49]: do we need a constructor to compile?

Comment (yotamdaniel) [reply contracts/Exchanger.sol:49]: We cant use a constructor for any contract that is upgradable (which all of those will be)

File: contracts/Coordinator.sol (hunk 14 above)
Comment (yotamdaniel) [reply contracts/Coordinator.sol:44]: We are using initialize() pattern but keeping the parameters of it to a minimum. When we deploy contracts, initialize() is the first thing that runs and is the constructor in a way. setDependencies is an admin action we'll run after we deploy. 

Makes deploying contracts simpler and less error prone and gives us more flexibility with Dependencies

File: contracts/Exchanger.sol (hunk 15 above)
Comment (yotamdaniel) [reply contracts/Exchanger.sol:49]: will remove commented out constructor though

File: contracts/Coordinator.sol (hunk 13 above)
Comment (yotamdaniel) [reply contracts/Coordinator.sol:16]: Good catch, thanks

File: contracts/AccessController.sol [hunk 16]
```diff
@@ -2,129 +2,81 @@
 pragma solidity 0.8.13;
 
 import {AccessControl} from "@openzeppelin/contracts/access/AccessControl.sol";
```
Comment (yotamdaniel) [inline contracts/AccessController.sol:4]: Try to remove

File: contracts/ParameterStore.sol [hunk 17]
```diff
@@ -3,14 +3,14 @@ pragma solidity 0.8.13;
...
+import {AccessController} from "./AccessController.sol";
 import "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
 import "@openzeppelin/contracts/utils/Strings.sol";
 
 /// @title ParameterStore is a contract for storing global parameters that can be modified by a privileged role
 /// @notice This contract (will be) proxy upgradable
-contract ParameterStore is AccessControl, Initializable, UUPSUpgradeable {
-    bytes32 public constant GOVERNOR_ROLE = keccak256("GOVERNOR_ROLE");
+contract ParameterStore is AccessController, UUPSUpgradeable {
+    // bytes32 public constant GOVERNOR_ROLE = keccak256("GOVERNOR_ROLE");
```
Comment (yotamdaniel) [inline contracts/ParameterStore.sol:13]: remove

File: contracts/PositionToken.sol [hunk 18]
```diff
@@ -3,46 +3,65 @@ pragma solidity 0.8.13;
 
 import "hardhat/console.sol";
 
-import {ERC721} from "@openzeppelin/contracts/token/ERC721/ERC721.sol";
-import {ERC721Burnable} from "@openzeppelin/contracts/token/ERC721/extensions/ERC721Burnable.sol";
-import {ERC721Enumerable} from "@openzeppelin/contracts/token/ERC721/extensions/ERC721Enumerable.sol";
-import {Counters} from "@openzeppelin/contracts/utils/Counters.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC721/ERC721Upgradeable.sol";
+// import "@openzeppelin/contracts-upgradeable/token/ERC721/ERC721EnumerableUpgradeable.sol";
```
Comment (yotamdaniel) [inline contracts/PositionToken.sol:7]: remove

File: test/ContractTestContext.ts [hunk 19]
```diff
@@ -64,15 +56,8 @@ async function deployContracts <T> (contractMap: ContractMap, contractRoles: Con
 
 type ArchContracts = {
     archToken: ArchToken;
-    cdp: CDPosition;
-    coordinator: Coordinator;
-    exchanger: Exchanger;
-    leverageAllocator: LeverageAllocator;
-    leverageEngine: LeverageEngine;
     lvUSD: LvUSDToken;
     // parameterStore: ParameterStore;
```
Comment (yotamdaniel) [inline test/ContractTestContext.ts:60]: remoce

**Review by jasonstockman — APPROVED** [review]

**Pull request merged**